server/src/main/java/org/elasticsearch/cluster/coordination/PeerFinder.java

```diff
@@ -0,0 +1,496 @@
+                        }
+                    }
+
+                    @Override
+                    public void onAfter() {
+                        super.onAfter();
+                        resolveInProgress.set(false);
+                    }
+
+                    @Override
+                    public String toString() {
+                        return "PeerFinder resolving unicast hosts list";
+                    }
+                });
+            }
+
+            final List<TransportAddress> transportAddresses = lastResolvedAddresses.get();
+            logger.trace("ActivePeerFinder#handleNextWakeUp(): probing supplied transport addresses {}", transportAddresses);
+            transportAddresses.forEach(this::startProbe);
+
+            futureExecutor.schedule(new Runnable() {
+                @Override
+                public void run() {
+                    handleWakeUp();
+                }
+
+                @Override
+                public String toString() {
+                    return "ActivePeerFinder::handleWakeUp";
+                }
+            }, findPeersDelay);
+        }
+
+        void startProbe(DiscoveryNode discoveryNode) {
+            startProbe(discoveryNode.getAddress());
+        }
+
+        void startProbe(TransportAddress transportAddress) {
+            assert holdsLock() : "PeerFinder mutex not held";
+            if (running == false) {
+                logger.trace("startProbe({}) not running", transportAddress);
+                return;
+            }
+
+            if (transportAddress.equals(getLocalNode().getAddress())) {
+                logger.trace("startProbe({}) not probing local node", transportAddress);
+                return;
+            }
+
+            if (probedAddressesSinceLastWakeUp.add(transportAddress) == false) {
+                logger.trace("startProbe({}) already probed this round", transportAddress);
+                return;
+            }
+
+            final DiscoveryNode cachedNode = connectedNodes.get(transportAddress);
+            if (cachedNode != null) {
+                assert cachedNode.getAddress().equals(transportAddress);
+
+                if (transportService.nodeConnected(cachedNode)) {
+                    logger.trace("startProbe({}) found connected {}", transportAddress, cachedNode);
+                    onProbeSuccess(cachedNode);
+                    return;
+                }
+
+                logger.trace("startProbe({}) found disconnected {}, probing again", transportAddress, cachedNode);
+                connectedNodes.remove(transportAddress, cachedNode);
+            } else {
+                logger.trace("startProbe({}) no cached node found, probing", transportAddress);
+            }
+
+            if (inFlightProbes.add(transportAddress)) {
+                logger.trace("startProbe({}) acquired probeLock", transportAddress);
+                executorService.get().execute(new AbstractRunnable() {
```

Comment:
There are two levels of async here that could arguably collapse to one, with `transportAddressConnector` doing its magic with its own `executorService` (possibly the same one).